File: Windows/Chronicy/Chronicy/Web/Converters/StackConverter.cs
using Chronicy.Data;
using System;
using System.Linq;

namespace Chronicy.Web.Converters
{
    public class StackConverter : IConverter<Data.Stack, Web.Models.Stack>
    {
        private Lazy<CustomFieldConverter> fieldConverter = new Lazy<CustomFieldConverter>();
        private Lazy<CardConverter> cardConverter = new Lazy<CardConverter>();

        public bool CanReverseConvert => true;

        public Models.Stack Convert(Data.Stack value)
        {
            return new Models.Stack
            {
                ID = value.ID,
                Name = value.Name,
                Fields = value.Fields.ConvertAll((input) => fieldConverter.Value.Convert(input)),
                Cards = value.Cards.ConvertAll(input => cardConverter.Value.Convert(input))
            };
        }

        public Data.Stack ReverseConvert(Models.Stack value)
        {
            return new Data.Stack(value.Name)
            {
                ID = value.ID,
                Fields = value.Fields.ConvertAll(input => fieldConverter.Value.ReverseConvert(input)),
                Cards = value.Cards.ConvertAll(input => cardConverter.Value.ReverseConvert(input))
            };
        }
    }

    public static class StackConverterExtensions
    {
        public static Models.Stack ToWebStack(this Data.Stack stack)
        {
            StackConverter converter = new StackConverter();
            return converter.Convert(stack);
        }

        public static Data.Stack ToDataStack(this Models.Stack stack)
        {
            StackConverter converter = new StackConverter();
            return converter.ReverseConvert(stack);
        }
    }
}

File: macOS/ChronicyFramework/Shared/Web/Converters/StackConverter.swift
//
//  StackConverter.swift
//  ChronicyFrameworkMacOS
//
//  Created by Alexandru Istrate on 25/07/2019.
//

import Foundation;

public class StackConverter: Converter<Stack, StackModel> {
    private let fieldConverter: CustomFieldConverter = CustomFieldConverter();
    private let cardConverter: CardConverter = CardConverter();
    
    public override var allowsReverseConvert: Bool { return true; }
    
    public override func convert(item: Stack) -> StackModel {
        let result: StackModel = StackModel();
//        result.id = item.id;
        result.name = item.name;
        result.fields = item.inputTemplate.fields.map({ (iter: CustomField) -> CustomFieldModel in
            return fieldConverter.convert(item: iter);
        });
        result.cards = item.cards.map({ (iter: Card) -> CardModel in
            return cardConverter.convert(item: iter);
        })
        return result;
    }
    
    public override func reverseConvert(item: StackModel) -> Stack {
        let result: Stack = Stack(name: item.name);
//        result.id = item.id;
        result.inputTemplate.fields = item.fields.map({ (iter: CustomFieldModel) -> CustomField in
            return fieldConverter.reverseConvert(item: iter);
        });
        result.cards = item.cards.map({ (iter: CardModel) -> Card in
            return cardConverter.reverseConvert(item: iter);
        })
        return result;
    }
}
